### ISSUE TITLE
Intermittent hanging when opening cached files without failure.

### ISSUE THREAD

**Issue by sharkinsspatial** [pangeo-forge/pangeo-forge-recipes#144]
In recent attempts at running the [SMAP SSS Recipe](https://github.com/pangeo-forge/staged-recipes/pull/31) we have noted intermittent hanging when opening from cache via [_maybe_open_or_copy_to_local](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/0.3.3/pangeo_forge_recipes/recipes/xarray_zarr.py#L61).  These are dask worker log entries from an recently attempted run where workers were hung

```
INFO:pangeo_forge_recipes.recipes.xarray_zarr:Opening 'https://podaac-opendap.jpl.nasa.gov/opendap/allData/smap/L3/RSS/V4/monthly/SCI/2015//RSS_smap_SSS_L3_monthly_2015_08_FNL_v04.0.nc' from cache
```
```
INFO:pangeo_forge_recipes.recipes.xarray_zarr:Opening 'https://podaac-opendap.jpl.nasa.gov/opendap/allData/smap/L3/RSS/V4/monthly/SCI/2015//RSS_smap_SSS_L3_monthly_2015_11_FNL_v04.0.nc' from cache
```
Interestingly, though the files which hang for cache retrieval are different between tests but consistently 3 workers seem to hang.  Not sure if this indicates some type of concurrency issue with cache retrieval.

Besides blocking the recipe, this is problematic due to the asynchronous nature of these operations which can potentially leave workers in an idle healthy state and prevent cleanup from the scheduler.

@rabernat I had look over the storage refactoring [PR](https://github.com/pangeo-forge/pangeo-forge-recipes/commit/32e920185934e42baf32650c5b3d1e76ea078ab3#diff-e12c886cc124886c5cfa5313d760a36c39649af9da845077c663e6feab8487b5L388) but I'm unsure if this might potentially address this issue?  Do you have recommendations on any additional logging or telemetry you would like me to collect that might be helpful in diagnosing the root cause for this?

If you feel the storage refactoring PR might address this issue, can we craft a new release which includes this so it can be deployed in our `staged-recipes` CI workflows.

**Comment by TomAugspurger** [comment]
Maybe cc @martindurant too. @sharkinsspatial are you able to provide the version of fsspec used by the workers? And do you know what the value of `copy_to_local` is in those runs?

**Comment by rabernat** [comment]
I don't think #133 would change this fundamentally. #133 moved some code around for organizational purposes. But all the i/o is still being handled by fsspec--that's where I would look for the root cause. Perhaps we could also turn on fsspec logging to see if it is emitting anything useful.

**Comment by rabernat** [comment]
Also noting that we still get occasional test hangs (e.g. https://github.com/pangeo-forge/pangeo-forge-recipes/pull/141/checks?check_run_id=2632512752). This was first documented in #33. This led to raising https://github.com/intake/filesystem_spec/issues/565, which was supposedly fixed by https://github.com/intake/filesystem_spec/pull/572. However, the issue was noted again more recently in #134.

There is something fundamentally unreliable going on here, and we need to get to the bottom of it.

**Comment by sharkinsspatial** [comment]
@TomAugspurger The `fsspec` used by the workers is `2021.4.0`.  With `XarrayZarrRecipe` the default for `copy_to_local` is `False` and this is not being overridden for [SMAP SSS Recipes](https://github.com/pangeo-forge/staged-recipes/pull/31/files#diff-65fa7571e642da92adb370cfd0780d4605a329b94127a4b390771cb101921673R183).

I'll enable `fsspec` logging at the `DEBUG` level and attempt another run to try and obtain some more details.

**Comment by sharkinsspatial** [comment]
I re-ran the recipe with `DEBUG` enabled for both `fsspec` and `s3fs` and encountered several hanging workers.  I made several runs and there is no notable consistency on which cached files the workers fail on reading.

Worker 1
```
[2021-05-25 17:56:24+0000] DEBUG - fsspec | <File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc> read: 54609 - 54617	1621965384798
DEBUG:fsspec:<File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc> read: 54609 - 54617	1621965384798
[2021-05-25 17:56:24+0000] DEBUG - s3fs | Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc, 54609-54617	1621965384798
DEBUG:s3fs:Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc, 54609-54617	1621965384798
[2021-05-25 17:56:24+0000] DEBUG - s3fs | CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc', 'Range': 'bytes=54609-54616', 'IfMatch': '"bd4245dd8fdf5069e42789eb1e284a34-3"'}	1621965384799
DEBUG:s3fs:CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc', 'Range': 'bytes=54609-54616', 'IfMatch': '"bd4245dd8fdf5069e42789eb1e284a34-3"'}	1621965384799
DEBUG:fsspec:<File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc> read: 45969 - 51729	1621965384830
[2021-05-25 17:56:24+0000] DEBUG - fsspec | <File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc> read: 45969 - 51729	1621965384830
DEBUG:s3fs:Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc, 45969-51729	1621965384830
[2021-05-25 17:56:24+0000] DEBUG - s3fs | Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc, 45969-51729	1621965384830
[2021-05-25 17:56:24+0000] DEBUG - s3fs | CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc', 'Range': 'bytes=45969-51728', 'IfMatch': '"bd4245dd8fdf5069e42789eb1e284a34-3"'}	1621965384830
DEBUG:s3fs:CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/82fe2b74-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2018_rss_smap_sss_l3_monthly_2018_10_fnl_v04.0.nc', 'Range': 'bytes=45969-51728', 'IfMatch': '"bd4245dd8fdf5069e42789eb1e284a34-3"'}	1621965384830
```

Worker 2
```
[2021-05-25 17:56:12+0000] DEBUG - fsspec | <File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc> read: 37691 - 38715	1621965372994
DEBUG:fsspec:<File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc> read: 37691 - 38715	1621965372994
[2021-05-25 17:56:12+0000] DEBUG - s3fs | Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc, 37691-38715	1621965372994
DEBUG:s3fs:Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc, 37691-38715	1621965372994
[2021-05-25 17:56:12+0000] DEBUG - s3fs | CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc', 'Range': 'bytes=37691-38714', 'IfMatch': '"a1a775e5a73794418336fcf574a86f2c-3"'}	1621965372995
DEBUG:s3fs:CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc', 'Range': 'bytes=37691-38714', 'IfMatch': '"a1a775e5a73794418336fcf574a86f2c-3"'}	1621965372995
[2021-05-25 17:56:13+0000] DEBUG - fsspec | <File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc> read: 30011 - 31035	1621965373015
DEBUG:fsspec:<File-like object S3FileSystem, pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc> read: 30011 - 31035	1621965373016
[2021-05-25 17:56:13+0000] DEBUG - s3fs | Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc, 30011-31035	1621965373016
DEBUG:s3fs:Fetch: pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl/NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc, 30011-31035	1621965373016
[2021-05-25 17:56:13+0000] DEBUG - s3fs | CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc', 'Range': 'bytes=30011-31034', 'IfMatch': '"a1a775e5a73794418336fcf574a86f2c-3"'}	1621965373016
DEBUG:s3fs:CALL: get_object - () - {'Bucket': 'pangeo-forge-aws-bakery-flowcachebucketpangeofor-196cpck7y0pbl', 'Key': 'NASA-SMAP-SSS/RSS/monthly/cache/813f2b66-https_podaac-opendap.jpl.nasa.gov_opendap_alldata_smap_l3_rss_v4_monthly_sci_2020_rss_smap_sss_l3_monthly_2020_01_fnl_v04.0.nc', 'Range': 'bytes=30011-31034', 'IfMatch': '"a1a775e5a73794418336fcf574a86f2c-3"'}	1621965373016
```

**Comment by sharkinsspatial** [comment]
ref https://github.com/dask/s3fs/issues/464

**Comment by martindurant** [comment]
You think fork might be an issue?

On May 25, 2021 6:06:05 PM EDT, Sean Harkins ***@***.***> wrote:
>ref https://github.com/dask/s3fs/issues/464
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/pangeo-forge/pangeo-forge-recipes/issues/144#issuecomment-848301313

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by sharkinsspatial** [comment]
@martindurant I'm unsure of dask's underlying process model on the workers but I was actually looking at the initial issue this thread which doesn't explicitly mention `fork` as the process creation method https://github.com/dask/s3fs/issues/464#issue-855523101.  This initial comment appears dask related and quite similar to what we are seeing.

**Comment by martindurant** [comment]
Please try with https://github.com/intake/filesystem_spec/pull/647 , which changes the instance caching logic to not keep instances alive in the cache itself.

**Comment by rabernat** [comment]
:+1: Trying that now in https://github.com/pangeo-forge/pangeo-forge-recipes/pull/146.

**Comment by rabernat** [comment]
No, it appears that intake/filesystem_spec#647 does not solve the problem. See https://github.com/pangeo-forge/pangeo-forge-recipes/pull/146/checks?check_run_id=2675850670

**Comment by rabernat** [comment]
Would it make sense if I refactored pangeo-forge-recipes to use aync for all its i/o and call `asyncio.get_running_loop()` to detect whether we are inside an existing event loop?

**Comment by martindurant** [comment]
worth a try!

The easiest thing to do would be to store whole files locally, then upload using get/put. You *cannot* share the loop with the sync API if you are already within the loop (the loop is not reentrant). 

Unfortunately, the file API is not easily async-able, because a file has a state of local position. It may be worthwhile implementing it, though, to enable the `f2.write(f.read(block))` pattern.

**Comment by martindurant** [comment]
> Copy them first to a local disk filesystem?

Exactly - or hold the whole file in memory (as opposed to on disc).
So you might have
```
await fs1._get_file(remote, temp_local)
await fs2._put_file(temp_local, remote)
```
or
```
data = await fs1._cat_file(remote)
await fs2._pipe_file(remote, data)
```
The former will be lighter on RAM. Obviously you need enough disc or RAM space to store whole files.

**Comment by rabernat** [comment]
The files we are dealing with will be up to 100 GB in size, so storing in RAM is a no-go.

As for
```
await fs1._get_file(remote, temp_local)
await fs2._put_file(temp_local, remote)
```
how is this different from copying from one fsspec filesystem to another? Rather than storing in `temp_local`, we want to store directly to remote? Passing through local disk is an unnecessary step I would really prefer to skip.

I will just copy the `_get_file` method and replace the local file with a remove write op.

**Comment by rabernat** [comment]
Btw, is there a reason there is no async implementation for the local filesystem?

https://github.com/intake/filesystem_spec/blob/master/fsspec/implementations/local.py

This will make it really hard to write generic code that works with s3, http, and local seamlessly.

**Comment by martindurant** [comment]
> is there a reason there is no async implementation for the local filesystem?

Yes, local files are inherently not really async. You can use `aiofile` or `aiofiles` (latter looks more prominent) or maybe others to emulate it via threads. Such approaches are nowhere near as standard as aiohttp and aiobotocore. Other filesystems like those depending on HTTP, (s)ftp, dask... could all be async given the effort.

**Comment by sharkinsspatial** [comment]
@martindurant Should we be treating these as these issues (hung tests potentially related to the use of `subprocess.Popen` and hanging within dask workers) as separate or the same potential failure case?  Internally `subprocess.Popen` is using [posix spawn](https://github.com/vstinner/cpython/blob/cc999ffbaaaa802872cf68457e51df2f8ec0676e/Lib/subprocess.py#L642) so based on this [feedback](https://github.com/dask/s3fs/issues/464#issuecomment-830846568) things should potentially be working here.  Or are both these issues likely more related to the interaction between `distributed`'s event loop and `fsspec`'s event loop?

In our call on Monday we discussed two potential options for mitigating this issue, refactoring recipes to use `asyncio` and [bringing](https://filesystem-spec.readthedocs.io/en/latest/async.html#bring-your-own-loop) `distributed`'s event loop to `fsspec`?  The lack of async for local filesystems seems to make refactoring recipes challenging.  @rabernat if @martindurant thinks it is worthwhile investigating would it be possible to update conftest to use `distributed`'s event loop?

**Comment by rabernat** [comment]
This issue is turning out to be quite difficult to pinpoint and therefore fix.

Can we produce a minimal reproducer? Something that can boil it down to the core essence?

The test that fails is `test_recipe_caching_copying[D-dask-False-False]`. Are we convinced that this is the same problem that @sharkinsspatial is experiencing in production?

In an attempt to boil it down further, in 40574a2e8ea48c6a01a502dc6809876a2db167d8 (part of #146) I am attempting to reproduce the error just via the `file_opener` function, which is where nearly all of the file copying logic now lives.

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/379e5bba1b64ed4011347a7bed82ed19b55a56dc/pangeo_forge_recipes/storage.py#L167-L201

We will see if that triggers a test failure.

Martin, if you have any insight into how to narrow down the search space and get to a minimal reproducer, it would be very appreciated.

**Comment by rabernat** [comment]
So I have managed to get more error logs out of github workflows: https://github.com/pangeo-forge/pangeo-forge-recipes/pull/146/checks?check_run_id=2678475789

The last line from the pangeo_forge_recipes logs is
```
Opening '/tmp/pytest-of-runner/pytest-0/netcdf_data1/000.nc' from cache
```

Which is essentially the same failure point as @sharkinsspatial 
```
INFO:pangeo_forge_recipes.recipes.xarray_zarr:Opening 'https://podaac-opendap.jpl.nasa.gov/opendap/allData/smap/L3/RSS/V4/monthly/SCI/2015//RSS_smap_SSS_L3_monthly_2015_11_FNL_v04.0.nc' from cache
```

What is interesting is that the test _only uses local files_. There is no s3, http, etc. And as we just learned, these fsspec filesystems don't use any async!

This is a pretty important clue. It suggests that the hanging is not really async-dependent. It's something else. Martin is this the right interpretation?

**Comment by martindurant** [comment]
That is a reasonable interpretation.

It is still conceptually possible for a hang to be caused by async/thread stuff from earlier, particularly if any objects get garbage collected (which happens in the main thread at some indeterminate time). 

If no fsspec async happened up to the failure point, then your interpretation is completely correct. The lack of the fsspec thread in the timeout stack says that this is indeed the case, at least for the main process we can see.

**Comment by rabernat** [comment]
@sharkinsspatial - I have been going nuts over in #146 trying to nail this down.

I have determined that if you set `copy_input_to_local_file=True`, the problem goes away--at least in the test suite.

Could you try this in your bakery environment?

If this works, it will be an acceptable workaround.

**Comment by sharkinsspatial** [comment]
🎊  @rabernat I ran a reduced dimension set of `noaa-oisst` recipe successfully with `copy_input_to_local_file=True` configured.  Once question here is if we have any heuristics on potential max required temporary storage?  In the AWS bakery, the workers are configured with a default of 20GB.  This is expandable to 200GB (which could be configured via a recipe's `meta.yaml`).

**Comment by rabernat** [comment]
> Once question here is if we have any heuristics on potential max required temporary storage?

The maximum size needed for temporary local storage will be determined by the input file size * number of input files per chunk. I think 20 GB is fine for now.

FWIW, I hope we don't have to keep this setting on permanently.

**Comment by rabernat** [comment]
FYI, we believe that @martindurant just fixed the root issue in #150.
